### PR TITLE
Cranelift/ISLE: allow extended left-hand-side patterns for clearer lowering rules

### PR DESCRIPTION
**Pull request by cfallin**
[Rendered](https://github.com/cfallin/rfcs/blob/isle-extended-patterns/accepted/isle-extended-patterns.md)

This RFC proposes an extension to the ISLE pattern-matching DSL used by Cranelift's code generation
backends. The extension is a backward-compatible proposal that is meant to make certain lowering
patterns easier to express. The ideas are informed by our experience so far with the DSL, and aim
to replace some awkward uses of extractors with in-arguments with clearer, more direct idioms
using a new language form.

cc @abrown @fitzgen @jlb6740 @akirilov-arm @sparker-arm @uweigand

### CODE REVIEW

**Comment by cfallin** [comment]
This is an elaboration of some of the high-level thoughts I mentioned in the [last Cranelift biweekly](https://github.com/bytecodealliance/wasmtime/blob/main/meetings/cranelift/2022/cranelift-04-18.md), and originally came from some discussions and head-scratching while pair-implementing some x64 rules with @abrown (thanks!). I spent some time playing with a bunch of different corners of the design space and this seems like the best option so far, but I'm curious if anyone has better ideas; I figured this was at the point that I could lay it out for feedback, in any case!

**Comment by sparker-arm** [comment]
This looks good to me... I'm currently trying to port IaddPairwise lowering to {U|S}ADDLP for aarch64 and I'm not finding it intuitive, so anything that makes it clearer how to use a N-to-1 extractor(s) would be great. My specific problem requires pattern matching a diamond shaped DAG and, I think, from your examples of current hacks, I can currently achieve this with using an approach from @uweigand. But, again, it doesn't look natural or really make obvious sense to me.

For my clarification, can current extractors use put_in_reg? Or is this considered as having side-effects? 

I also like having the 'given' for predicates, such as for architecture features.

**Comment by cfallin** [comment]
Thanks! For what it's worth @sparker-arm I'd be happy to help you work out how to get particular lowerings to work sometime, if you think that would be useful. (I've done some Zoom pair-implementing with abrown before and it's always interesting and educational for both of us!)

> For my clarification, can current extractors use put_in_reg? Or is this considered as having side-effects?

It's a side-effecting constructor, so this would only come in the RHS. In general the phasing divide that makes sense I think is to collect as much as we need to match in the LHS (this includes "pure" parts that we may or may not be able to build, like certain immediate kinds), and then start the emission actions in the RHS. The only kinds of actions that *have* to come in the RHS (after we commit) are `put_in_reg`, `sink_inst`, and `emit`, I think.

**Comment by cfallin** [comment]
Ah, I forgot to cc @avanhatt re: verification concerns -- please let us know what you think as well!

**Comment by fitzgen** [comment]
This example

```lisp
(rule (lower (magic_simd_op x y))
      (given (magic_simd_extension_enabled))
      (x64_magic_simd_op x y))
```

seems like it would be better written as the following so that it isn't "backwards" -- which users have found confusing -- with the new language extension:

```lisp
(rule (lower (magic_simd_op x y))
      (given _ (magic_simd_extension_enabled))
      (x64_magic_simd_op x y))
```

**Comment by fitzgen** [comment]
Overall, I like this a lot!

**Comment by fitzgen** [comment]
Are you imagining that pure external terms would only get `&self` context methods?

**Comment by cfallin** [comment]
> (And that would let us get rid of optional expressions in `given` clauses, simplifying the language extension a little)

Hmm, yeah, I like your version better here, thanks!

**Comment by sparker-arm** [comment]
> I'd be happy to help you work out how to get particular lowerings to work sometime

Thanks!

> It's a side-effecting constructor, so this would only come in the RHS.

Sorry to go off track, but why is it that to look at a value produced by an instruction, we have to put_in_reg first? Can't extractors understand an SSA value from an instruction, without causing side-effects? As a concrete example, is something like a simple optimisation like this possible?

```
(rule (lower (isub x y))
  (given (are_equal x  y))
  (iconst 0))
```

File: accepted/isle-extended-patterns.md
Comment (avanhatt) [inline accepted/isle-extended-patterns.md:202]: Can we expand on this a bit? I.e., if any `given` clause is unable to match, the rule does not match? Also, including the case of a boolean predicate more explicitly here?

File: accepted/isle-extended-patterns.md
Comment (avanhatt) [inline accepted/isle-extended-patterns.md:205]: The "with no `given` clauses" bit here doesn't make sense to me---presumably, the elided RHS here uses variables bound in the `given` clauses, so what does it mean to behave the same as before?

File: accepted/isle-extended-patterns.md
Comment (avanhatt) [inline accepted/isle-extended-patterns.md:210]: To this end, would it make sense to name this something like `if-let` instead of `given`? I ask both for clarity in my own understanding and as an actual naming suggestion :)

File: accepted/isle-extended-patterns.md
Comment (avanhatt) [inline accepted/isle-extended-patterns.md:217]: Might be helpful here to explicitly call out the difference between RHS expressions and LHS patterns again, since "expression" could be read in a less-specific way.

**Comment by cfallin** [comment]
> Sorry to go off track, but why is it that to look at a value produced by an instruction, we have to put_in_reg first? Can't extractors understand an SSA value from an instruction, without causing side-effects? As a concrete example, is something like a simple optimisation like this possible?
> 
> ```
> (rule (lower (isub x y))
>   (given (are_equal x  y))
>   (iconst 0))
> ```

That's definitely possible! `put_in_reg` isn't needed to look at the SSA `Value`; it is only needed once one decides to put the value in a register and use it in a machine instruction.

File: accepted/isle-extended-patterns.md
Comment (avanhatt) [reply accepted/isle-extended-patterns.md:205]: Makes sense! The singular `the rule` was what tripped me, maybe better as `(rules without `given` clauses behave the same as before)`

File: accepted/isle-extended-patterns.md
Comment (fitzgen) [reply accepted/isle-extended-patterns.md:210]: I like `if-let` instead of `given`

Comment (cfallin) [reply accepted/isle-extended-patterns.md]: Alright, I've gone ahead and find-replaced everything to `if-let` instead. Thanks for the idea!

File: accepted/isle-extended-patterns.md
Comment (cfallin) [reply accepted/isle-extended-patterns.md]: Fixed, hopefully clearer now.

File: accepted/isle-extended-patterns.md
Comment (cfallin) [reply accepted/isle-extended-patterns.md]: Updated, hopefully more descriptive now.

File: accepted/isle-extended-patterns.md
Comment (cfallin) [reply accepted/isle-extended-patterns.md]: Expanded this with a more explicit discussion of how the evaluation proceeds.

**Comment by cfallin** [comment]
Thanks all -- updated based on feedback!

It seems like there are no really big concerns here yet (and after speaking with @avanhatt, @mlfbrown and @mpardesh today it sounds like this should work for `isle-veri` as well), but please let me know if there are any other lingering concerns. Otherwise I'll likely move this to final-call status in a day or two.

**Comment by cfallin** [comment]
Alright, it is now time for the FCP, I think! I'll create the stakeholders list below based on committers to Cranelift active in the past three months (since Jan 26, anything more than a trivial doc fix) and Cranelift-related RFCs, plus anyone who has commented here.

# Motion to finalize with a disposition to merge

# Stakeholders sign-off

## Arm

* [ ] @akirilov-arm
* [ ] @sparker-arm
* [ ] @dheaton-arm
* [ ] @FreddieLiardet

## Cornell

* [ ] @avanhatt

## Fastly

* [ ] @alexcrichton 
* [x] @cfallin
* [ ] @fitzgen 

## IBM

* [ ] @uweigand

## Intel

* [ ] @abrown 
* [ ] @jlb6740

## Unaffiliated

* [ ] @bjorn3

**Comment by cfallin** [comment]
With approvals from two separate groups (Fastly and Intel) this now enters the *Final Comment Period*, which will end on Mon May 9 (unless all groups have at least one approval, shortcutting the process, or any objections are raised, pausing it).

**Comment by uweigand** [comment]
Looks good to me, thanks!

**Comment by sparker-arm** [comment]
Yeah, LGTM too, thanks!

File: accepted/isle-extended-patterns.md
Comment (bjorn3) [inline accepted/isle-extended-patterns.md:302]: Maybe this should just be `(if (magic_simd_extension_enabled))`?

**Review by bjorn3 — APPROVED** [review]

**Comment by cfallin** [comment]
With approvals from all stakeholder groups above, the final comment period has ended early and this RFC will merge:  as per the [RFC process](https://github.com/bytecodealliance/rfcs/blob/main/accepted/rfc-process.md),

> Finally, the RFC is automatically merged/close if either:
> * The FCP elapses without any objections.
> * A stakeholder from each group has signed off, short-cutting the waiting period.

Thanks all! I'll work on cleaning up the draft implementation in bytecodealliance/wasmtime#4072, incorporating changes from here (notably `given` -> `if-let`), and hopefully we can start using this in our lowering rules soon!

**Pull request merged**
